lighthouse-core/lib/client-rect-functions.js

```diff
@@ -0,0 +1,311 @@
+      if (possiblyContainingRect === cr) {
+        continue;
+      }
+      if (filteredOutRects.has(possiblyContainingRect)) {
+        // Can't contain anything because we removed it from the list
+        continue;
+      }
+      if (rectContains(possiblyContainingRect, cr)) {
+        filteredOutRects.add(cr);
+        return false;
+      }
+    }
+    return true;
+  });
+}
+
+/**
+ * @param {number} a
+ * @param {number} b
+ */
+function almostEqual(a, b) {
+  // Sometimes a child will reach out of the parent by
+  // 1px or 2px, so be somewhat tolerant for merging
+  return Math.abs(a - b) <= 2;
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect} rect
+ */
+function getRectCenterPoint(rect) {
+  return {
+    x: rect.left + rect.width / 2,
+    y: rect.top + rect.height / 2,
+  };
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect} crA
+ * @param {LH.Artifacts.ClientRect} crB
+ * @returns {boolean}
+ */
+function clientRectsTouch(crA, crB) {
+  // https://stackoverflow.com/questions/2752349/fast-rectangle-to-rectangle-intersection
+  return (
+    crA.left <= crB.right &&
+    crB.left <= crA.right &&
+    crA.top <= crB.bottom &&
+    crB.top <= crA.bottom
+  );
+}
+
+/**
+ * @param {LH.Artifacts.ClientRect[]} clientRects
+ * @returns {LH.Artifacts.ClientRect[]}
+ */
+function mergeTouchingClientRects(clientRects) {
+  for (let i = 0; i < clientRects.length; i++) {
+    for (let j = i + 1; j < clientRects.length; j++) {
+      const crA = clientRects[i];
+      const crB = clientRects[j];
+
+      /**
+       * Examples of what we want to merge:
+       *
+       * AAABBB
+       *
+       * AAA
+       * AAA
+       * BBBBB
+       */
+      const rectsLineUpHorizontally =
+        almostEqual(crA.top, crB.top) || almostEqual(crA.bottom, crB.bottom);
+      const rectsLineUpVertically =
+        almostEqual(crA.left, crB.left) || almostEqual(crA.right, crB.right);
+      const canMerge =
+        clientRectsTouch(crA, crB) &&
+        (rectsLineUpHorizontally || rectsLineUpVertically);
+
+
+      if (canMerge) {
+        // create rect that contains both crA and crB
+        const left = Math.min(crA.left, crB.left);
+        const right = Math.max(crA.right, crB.right);
+        const top = Math.min(crA.top, crB.top);
+        const bottom = Math.max(crA.bottom, crB.bottom);
+        const replacementClientRect = addRectWidthAndHeight({
+          left,
+          right,
+          top,
+          bottom,
+        });
+
+        const mergedRectCenter = getRectCenterPoint(replacementClientRect);
+        if (
+          !(
+            rectContainsPoint(crA, mergedRectCenter) ||
+            rectContainsPoint(crB, mergedRectCenter)
+          )
+        ) {
+          // Don't merge because the new shape is too different from the
+          // merged rects, and tapping in the middle wouldn't actually hit
+          // either rect
+          continue;
+        }
+
+        clientRects = clientRects.filter(cr => cr !== crA && cr !== crB);
```

Comment:
maybe just a quick comment that we're replacing the two client rects with the merged version? the recursive call is a bit unexpected too, I'm guessing it was messy to deal with concurrent modification but a bit of explanation of heads up in function description would help

Reply:
> the recursive call is a bit unexpected too

Yeah it felt more simpler to just start over than modifying the indices and checking if the mutation introduces any edge cases. For client rects the arrays are also rarely more than 5 rects, so starting again doesn't cause perf issues.

Reply:
> For client rects the arrays are also rarely more than 5 rects, so starting again doesn't cause perf issues.

👍 throwing that in a comment would also be ace :)